grudge/trace_pair.py

```diff
@@ -272,61 +277,130 @@ def connected_ranks(dcoll: DiscretizationCollection):
     return get_connected_partitions(dcoll._volume_discr.mesh)
 
 
+def _flatten_to_numpy_from_container(
+        actx: ArrayContext, ary: ArrayOrContainerT) -> ArrayT:
+    """Flattens an arbitrary array container *ary* into a one-dimensional
+    numpy array suitable for communication via mpi4py.
+    """
+    if is_array_container(ary):
+        flat_data = []
+        # NOTE: The ordering of data in the flattened array will be determined
+        # by the ordering of key/value pairs returned by *serialize_container*.
+        # This is important to remember because the function
+        # *_unflatten_from_numpy_to_container* relies on this same ordering
+        # to unflatten the concatenated array.
+        for _, value in serialize_container(to_numpy(ary, actx)):
+            # NOTE: *to_numpy* can return numpy object arrays whose components
+            # are numpy arrays. As a result, we need to flatten these as well.
+            if isinstance(value, np.ndarray) and value.dtype == "O":
+                flat_data.append(np.concatenate(value))
+            else:
+                flat_data.append(value)
+        # All data is concatenated into a single numpy array
+        flattened_np_ary = np.concatenate(flat_data)
+    else:
+        flattened_np_ary = flatten_to_numpy(actx, ary)
+    return flattened_np_ary
+
+
+def _unflatten_from_numpy_to_container(
+        actx: ArrayContext,
+        template: ArrayOrContainerT,
+        numpy_ary, discr) -> ArrayOrContainerT:
+    """Unflattens *numpy_ary* into an array container matching
+    the provided container template *template*.
+    """
+    if not isinstance(template, DOFArray):
+        container_data = []
+        offset = 0
+        # NOTE: The flattened data in *numpy_ary* is assumed to have
+        # the same ordering determined by the key/pair-ordering
+        # of *serialize_container(template)*.
+        for key, _value in serialize_container(template):
+            # NOTE: The values of *template* determine how the values of
+            # the re-containerized data should look. For example, if we need
+            # to create an object array as one of the components.
+            # To construct the object array from a slice of flattened data,
+            # we need to inspect *_value* from *template* to determine
+            # the right offset parameter for accessing *numpy_ary*. The actual
+            # data in *_value* is not needed.
+            if isinstance(_value, np.ndarray) and _value.dtype == "O":
```

Comment:
Why are object arrays handled specially here? Can't they go through the same logic as all other containers? (If not, add a comment to explain.)

Reply:
@inducer Because serialize_container does not serialize a nested object array inside an array container and the flattened array has absolutely no clue what array container it originally came from. It's basically what I wrote in the `# NOTE: ...`